### ISSUE TITLE
'mps' and argmax()

### ISSUE THREAD

**Issue by dbl001** [pytorch/pytorch#92311]
### 🐛 Describe the bug

tensor.argmax() returning erroneous result on 'mps' backend.
```
% ipython
Python 3.10.8 (main, Nov 24 2022, 08:09:04) [Clang 14.0.6 ]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.7.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import torch

In [2]: a = torch.randn(10,2, device='mps', dtype=torch.float32)

In [3]: a
Out[3]: 
tensor([[ 1.3196, -0.2676],
        [ 0.6754, -1.3163],
        [-0.2589,  1.2122],
        [-0.2729, -0.2693],
        [-0.8443, -0.4162],
        [ 0.6202, -0.2219],
        [ 1.3816,  0.8552],
        [ 0.1368, -0.0287],
        [ 0.0820, -1.4613],
        [-1.1522,  0.2325]], device='mps:0')

In [4]: torch.argmax(a, dim=1)
Out[4]: 
tensor([-9223372036854775808, -9223372036854775808, -9223372036854775808,
        -9223372036854775808, -9223372036854775808, -9223372036854775808,
        -9223372036854775808, -9223372036854775808, -9223372036854775808,
        -9223372036854775808], device='mps:0')

In [5]: b = torch.randn(10,2, device='cpu', dtype=torch.float32)

In [6]: b
Out[6]: 
tensor([[ 0.2422,  0.5932],
        [-0.0680,  0.0885],
        [-1.6327, -0.1222],
        [ 0.8754,  1.2478],
        [ 2.0334, -2.3631],
        [ 0.0316, -0.1020],
        [ 1.6928,  1.2704],
        [-1.2770,  0.3706],
        [-1.0360,  0.5607],
        [ 0.6821,  0.6692]])

In [7]: torch.argmax(b, dim=1)
Out[7]: tensor([1, 1, 1, 1, 0, 0, 0, 1, 1, 0])


```

### Versions

```
% python collect_env.py
Collecting environment information...
PyTorch version: 2.0.0a0+gitf4b804e
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.1 (x86_64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: version 3.22.1
Libc version: N/A

Python version: 3.10.8 (main, Nov 24 2022, 08:09:04) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-10.16-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] audiolm-pytorch==0.0.1
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.5
[pip3] pytorch-transformers==1.1.0
[pip3] torch==2.0.0a0+gitf8b2879
[pip3] torch-struct==0.5
[pip3] torch-summary==1.4.5
[pip3] torchaudio==0.13.0.dev20221015
[pip3] torchtraining-nightly==1604016577
[pip3] torchvision==0.15.0a0+8985b59
[pip3] vector-quantize-pytorch==0.9.2
[conda] nomkl                     3.0                           0  
[conda] numpy                     1.23.5          py310he50c29a_0  
[conda] numpy-base                1.23.5          py310h992e150_0  
[conda] pytorch-transformers      1.1.0                    pypi_0    pypi
[conda] torch                     2.0.0a0+gitdcb73aa          pypi_0    pypi
[conda] torch-struct              0.5                      pypi_0    pypi
[conda] torch-summary             1.4.5                    pypi_0    pypi
[conda] torchaudio                0.13.0.dev20221015          pypi_0    pypi
[conda] torchtraining-nightly     1604016577               pypi_0    pypi
[conda] torchvision               0.15.0a0+8985b59          pypi_0    pypi
[conda] vector-quantize-pytorch   0.9.2                    pypi_0    pypi
```

cc @ezyang @gchanan @zou3519 @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

**Comment by JoseMiguelCh** [comment]
Same issue on Mac M1, with PyTorch '2.0.0.dev20230205' and Python 3.10.9

**Comment by clchrlls** [comment]
Same issue on x86 with 'pytorch 2.0.0.dev20230226' and python 3.9.16, but I found a workaround. 

Just replace` argmax(...)` with `max(...).indices` for instance replace `output.argmax(dim=1)` with `output.max(dim=1).indices`

**Comment by nshervt** [comment]
Same issue:

> ```
> Collecting environment information...
> PyTorch version: 2.0.0.dev20230227
> Is debug build: False
> CUDA used to build PyTorch: None
> ROCM used to build PyTorch: N/A
> 
> OS: macOS 13.2.1 (x86_64)
> GCC version: Could not collect
> Clang version: 14.0.0 (clang-1400.0.29.202)
> CMake version: Could not collect
> Libc version: N/A
> 
> Python version: 3.8.16 | packaged by conda-forge | (default, Feb  1 2023, 16:05:36)  [Clang 14.0.6 ] (64-bit runtime)
> Python platform: macOS-10.16-x86_64-i386-64bit
> Is CUDA available: False
> CUDA runtime version: No CUDA
> CUDA_MODULE_LOADING set to: N/A
> GPU models and configuration: No CUDA
> Nvidia driver version: No CUDA
> cuDNN version: No CUDA
> HIP runtime version: N/A
> MIOpen runtime version: N/A
> Is XNNPACK available: True
> 
> CPU:
> Apple M1
> 
> Versions of relevant libraries:
> [pip3] numpy==1.24.2
> [pip3] torch==2.0.0.dev20230227
> [pip3] torchvision==0.15.0.dev20230227
> [conda] numpy                     1.24.2                   pypi_0    pypi
> [conda] torch                     2.0.0.dev20230227          pypi_0    pypi
> [conda] torchvision               0.15.0.dev20230227          pypi_0    pypi
> ```

Sample code:

```
import torch
import torch.nn.functional as F

device = torch.device('mps')
logits = torch.tensor([0.5749, -0.0438, -0.0700, -0.1062, -0.0332, -0.0981, -0.0614, -0.0751, -0.0396, -0.0733], device=device)
pred = F.softmax(logits).argmax()

print(pred)
```

Output:

> ```
> tensor(-9223372036854775808, device='mps:0')
> 
> Process finished with exit code 0
> ```

**Comment by TheBloke** [comment]
This issue is preventing me from loading any LLMs via Transformers on my Intel system with AMD 6900XT GPU.

Happens in 2.0.0 and also 2.1.0.dev20230402.

Any fix would be much appreciated.

**Comment by msaroufim** [comment]
cc @kulinseth - this is breaking all HF model per @sgugger

**Comment by sgugger** [comment]
The generate function on all text models to be more specific.

**Comment by kulinseth** [comment]
Thanks @sgugger , we are taking a look.

**Comment by kulinseth** [comment]
@sgugger , I tried on M1 machine with macOS 13.3 and torch 2.0 
```
>>> import torch
>>> a = torch.randn(10,2, device='mps', dtype=torch.float32)
>>> a
tensor([[-0.7468,  0.7802],
        [-1.0569, -0.6632],
        [-0.0833, -0.4993],
        [ 0.5678, -0.9434],
        [ 0.5763,  0.7651],
        [-1.0167,  0.4784],
        [ 1.3732, -1.1688],
        [ 0.7816, -0.0552],
        [-0.4203, -0.0046],
        [-1.1244,  0.5733]], device='mps:0')
>>> torch.argmax(a, dim=1)
tensor([1, 1, 0, 0, 1, 1, 0, 0, 1, 1], device='mps:0')
```
And got expected results. Can you please update to MacOS 13.3 and check?

**Comment by dbl001** [comment]
On MacOS 13.3.1. Just pulled and compiled 'pytorch' from Master. GPU: AMD Radeon Pro 5700 XT
```
% ipython
Python 3.10.10 (main, Mar 21 2023, 13:41:39) [Clang 14.0.6 ]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.12.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import torch
In [3]: a =  torch.randn(10,2, device='mps', dtype=torch.float32)

In [4]: a
Out[4]: 
tensor([[ 0.1299,  0.7419],
        [ 0.7063, -0.8204],
        [-0.9533,  0.5050],
        [-1.4201,  1.5996],
        [ 0.9593, -1.2127],
        [ 0.3628, -0.6628],
        [-0.3352, -1.9211],
        [ 0.2966, -2.0903],
        [-0.6901,  0.4967],
        [ 0.5417,  0.0655]], device='mps:0')

In [5]: torch.argmax(a, dim=1)
Out[5]: 
tensor([-9223372036854775808, -9223372036854775808, -9223372036854775808,
        -9223372036854775808, -9223372036854775808, -9223372036854775808,
        -9223372036854775808, -9223372036854775808, -9223372036854775808,
        -9223372036854775808], device='mps:0')

In [7]: print(torch.__version__)
2.1.0.dev20230406
```

**Comment by TheBloke** [comment]
> And got expected results. Can you please update to MacOS 13.3 and check?

Yes I think this might an Intel/AMD specific issue then. Here's my results running the same:
```
In [1]: import torch
In [2]: a = torch.randn(10,2, device='mps', dtype=torch.float32)
In [3]: a
Out[3]: tensor([[-0.2422,  1.8948],
        [ 0.7428, -0.2871],
        [-0.9586, -0.1957],
        [-1.2639,  1.8659],
        [ 0.2276,  0.5398],
        [ 1.1572, -0.1739],
        [-1.7101,  0.2622],
        [-1.1599, -2.1204],
        [-0.3746, -0.5579],
        [ 0.6552,  0.6651]], device='mps:0')
In [4]: torch.argmax(a, dim=1)
Out[4]:
tensor([-9223372036854775808, -9223372036854775808, -9223372036854775808,
        -9223372036854775808, -9223372036854775808, -9223372036854775808,
        -9223372036854775808, -9223372036854775808, -9223372036854775808,
        -9223372036854775808], device='mps:0')
```

And the same problem applies to `torch.multinomial`:
```
In [2]: c
Out[2]:
tensor([[0.1000],
        [0.2000],
        [0.3000],
        [0.4000]])
In [3]: c = c.to('mps')
In [4]: torch.multinomial(c, num_samples=1)
Out[4]:
tensor([[-9223372036854775808],
        [-9223372036854775808],
        [-9223372036854775808],
        [-9223372036854775808]], device='mps:0')
```
I am running Ventura 13.3 on an Intel system with AMD 6900XT GPU:

`Darwin Eddie.lan 22.4.0 Darwin Kernel Version 22.4.0: Mon Mar  6 21:00:17 PST 2023; root:xnu-8796.101.5~3/RELEASE_X86_64 x86_64`

<img width="280" alt="image" src="https://user-images.githubusercontent.com/784313/230734628-93151f87-be79-4026-9974-0b98d885f6dd.png">

**Comment by sb1992** [comment]
Any updates.. Still not working, i reckon..

**Comment by TheBloke** [comment]
Yeah it'd be really nice to get some update here.  Is it a PyTorch bug, or is it a macOS issue?

I realise that Intel macOS + AMD GPU is not the future, but there are millions of users who still have these systems and would love to be able to try MPS locally.

**Comment by notSaranshMalik** [comment]
Confirming I got the same error on an M1 mac

**Comment by Guanzhou-Ke** [comment]
I also found this issue in MBP'17 (Intel + AMD Radeon Pro 560). It is working well when I replace `argmax(dim=1)` to `max(dim=1).indices`.

**Comment by DenisVieriu97** [comment]
Fixed in https://github.com/pytorch/pytorch/pull/104374

**Comment by dbl001** [comment]
```
 % ipython
Python 3.10.11 (main, Apr 20 2023, 13:59:00) [Clang 14.0.6 ]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.12.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import torch

In [2]: a = torch.randn(10,2, device='mps', dtype=torch.float32)

In [3]: torch.argmax(a, dim=1)
Out[3]: 
tensor([-9223372036854775808, -9223372036854775808, -9223372036854775808,
        -9223372036854775808, -9223372036854775808, -9223372036854775808,
        -9223372036854775808, -9223372036854775808, -9223372036854775808,
        -9223372036854775808], device='mps:0')

In [4]: quit
(AI-Feynman) davidlaxer@bluediamond pix2struct % print(torch.__version__)
zsh: unknown sort specifier
(AI-Feynman) davidlaxer@bluediamond pix2struct % ipython                 
Python 3.10.11 (main, Apr 20 2023, 13:59:00) [Clang 14.0.6 ]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.12.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import torch

In [2]: print(torch.__version__)
2.1.0.dev20230630

In [3]: 

```
I'm trying to build from source to see if it's working.  Stay tuned.

**Comment by dbl001** [comment]
I works when I build from source:
```
% ipython 
Python 3.10.11 (main, Apr 20 2023, 13:59:00) [Clang 14.0.6 ]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.12.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import torch

In [2]: print(torch.__version__)
2.1.0a0+git28720ad

In [3]: a = torch.randn(10,2, device='mps', dtype=torch.float32)

In [4]: torch.argmax(a, dim=1)
Out[4]: tensor([0, 0, 1, 1, 0, 1, 0, 0, 1, 1], device='mps:0')

```
Thanks!

**Comment by TheBloke** [comment]
Thank you! 

Does this fix the `torch.multinomial` issue at the same time?

**Comment by DenisVieriu97** [comment]
> Thank you!
> 
> Is this expected to fix the `torch.multinomial` issue at the same time?

Yes, this will fix the multinomial issue too.
